### PR TITLE
fail if [kill] contains [filter], re #4316

### PR DESCRIPTION
**Pull request by ProditorMagnus**

### CODE REVIEW

**Comment by soliton-** [comment]
The question is how useful it is to add arbitrary amounts of such checks everywhere. This should be caught by validating the WML, no?

Additionally we could support [filter] everywhere where it is inlined now and avoid the confusion altogether? Inlining the filter is a bad design anyway since it leads to key name collisions and it'd be nice to move away from this.

**Comment by ProditorMagnus** [comment]
I agree, but I am not able to provide full solution for #4316 and it has been long enough that I felt even partial solution would help.

**Comment by gfgtdf** [comment]
Alternatively we could just allow filter there, then we'd have some line `local filter = wml.get_child(cfg, "filter) or cfg`and use the variable filter later.

**Comment by CelticMinstrel** [comment]
> This should be caught by validating the WML, no?

There is no expectation that add-on WML will be validated.

**Comment by stevecotton** [comment]
Given that an add-on might already have been tested and rebalanced without spotting the bug, would it be better to show a warning instead of a hard error? Possibly not, I'm just wondering aloud.

**Comment by ProditorMagnus** [comment]
I followed the idea of [modify_unit] - there [filter] is mandatory, even if I want to change all units.

**Comment by cooljeanius** [comment]
I can confirm that I've been bitten by the problem that this PR is addressing before, although I wonder if this is the correct place to address it... maybe a `wmllint` check instead?

**Comment by ProditorMagnus** [comment]
I believe previous answer `There is no expectation that add-on WML will be validated` fits wmllint too.

This PR is not the greatest solution, but I believe it still brings value.

**Comment by CelticMinstrel** [comment]
I think putting it here is definitely better than putting it in wmllint or the schema.

Perhaps an even better place would be in the filter parsers (in C++) – if they see a `[filter]` tag (or, perhaps, any tag whose name begins with "filter") they should complain that it's invalid. That would mean you automatically get the warning for _any_ tag that uses an embedded filter if you add any filter tag to it.

**Comment by cooljeanius** [comment]
> I think putting it here is definitely better than putting it in wmllint or the schema.
> 

All right, how about if we say "in addition to" instead of "instead"?

File: data/lua/wml/kill.lua [hunk 1]
```diff
@@ -3,6 +3,10 @@ local location_set = wesnoth.require "location_set"
 local kill_recursion_preventer = location_set.create()
 
 function wesnoth.wml_actions.kill(cfg)
+	if wml.get_child(cfg, "filter") then
+		wml.error "do not use [filter] in [kill]"
```
Comment (stevecotton) [inline data/lua/wml/kill.lua:7]: The writing style doesn't match similar error messages that are already in the code. Suggestions:

* "[kill] has unsupported [filter] subtag"
* "[filter] tag found inside [kill]"

Comment (ProditorMagnus) [reply data/lua/wml/kill.lua:7]: `"[kill] has unsupported [filter] subtag"` looks nicer.

**Comment by CelticMinstrel** [comment]
> I think putting it here is definitely better than putting it in wmllint or the schema.
> 
> Perhaps an even better place would be in the filter parsers (in C++) – if they see a `[filter]` tag (or, perhaps, any [i]unknown[/i] tag whose name begins with "filter") they should complain that it's invalid. That would mean you automatically get the warning for _any_ tag that uses an embedded filter if you add any filter tag to it. The only potential downside to that is that I'm not sure if that'll give you the identity of the tag that add the incorrect filter in it… I guess it would probably show up in a stack trace if nothing else?

File: data/lua/wml/kill.lua (hunk 1 above)
Comment (ProditorMagnus) [reply data/lua/wml/kill.lua:7]: schema uses 20230713 19:25:15 error validation: Tag [event] may not be used in [effect]

Comment (stevecotton) [reply data/lua/wml/kill.lua:7]: That sounds good to me too.

**Review by stevecotton — APPROVED** [review]

**Pull request merged**